File: 2709-Greatest Common Divisor Traversal.py
class Solution:
    def canTraverseAllPairs(self, nums: List[int]) -> bool:
        if len(nums) == 1:
            return True
        
        if 1 in nums:
            return False
        
        def find(x):
            if parent[x] != x:
                parent[x] = find(parent[x])
            return parent[x]
        
        def union(a, b):
            pa, pb = find(a), find(b)
            if pa != pb:
                parent[pb] = pa
        
        nums = list(set(nums))
        N = len(nums)
        parent = {i:i for i in range(N)}
        factorMap= {}

        for i, num in enumerate(nums):
            for j in range(2, isqrt(num) + 1):
                if num % j == 0:
                    if j not in factorMap:
                        factorMap[j] = i
                    else:
                        union(i, factorMap[j])
                    while num%j == 0:
                        num = num//j
            if num == 1:
                continue
            if num not in factorMap:
                factorMap[num] = i
            else:
                union(i, factorMap[num])
                
        connected = set()
        for i in range(N):
            connected.add(find(i))

        return len(connected) == 1

File: C++/2709-Greatest Common Divisor Traversal.cpp
vector<int> SieveOfEratosthenes(int n) {
    vector<bool> prime(n+1, true);
    prime[0] = prime[1] = false;
    for (int p=2; p*p<=n; p++) {
        if (prime[p] == true) {
            for (int i=p*p; i<=n; i += p)
                prime[i] = false;
        }
    }
    vector<int> res;
    for (int p=2; p<=n; p++)
       if (prime[p])
           res.push_back(p);
    return res;
}

vector<int> primes = SieveOfEratosthenes(100000);

class Solution {
public:
    int find(int x, vector<int>& parent) {
        if (x != parent[x]) {
            parent[x] = find(parent[x], parent);
        }
        return parent[x];
    }
    
    void uni(int a, int b, vector<int>& parent) {
        int pa = find(a, parent), pb = find(b, parent);
        if (pa != pb) parent[pa] = pb;
    }
    
    bool canTraverseAllPairs(vector<int>& nums) {
        int N = nums.size();
        unordered_map<int, vector<int>> hmap;
        vector<int> parent(N);
        for (int i=0; i<N; i++) {
            parent[i] = i;
        }
        
        for (int i=0; i<N; i++) {
            int tmp = nums[i];
            for (int p:primes) {
                if (p*p > tmp) break;
                if (tmp%p == 0) {
                    while (tmp%p == 0) tmp = tmp/p;
                    hmap[p].push_back(i);
                }
            }
            if (tmp == 1) continue;
            hmap[tmp].push_back(i);
        }
        
        for (auto &kv:hmap) {
            for (int i=1; i<kv.second.size(); i++) {
                uni(kv.second[i-1], kv.second[i], parent);
            }
        }

        for (int i=1; i<N; i++) {
            if (find(i-1, parent) != find(i, parent)) {
                return false;
            }
        }
        
        return true;
    }
};